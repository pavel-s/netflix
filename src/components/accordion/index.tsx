import {
  createContext,
  Dispatch,
  SetStateAction,
  useContext,
  useState,
} from 'react';
import { TRSC } from '../../types';
import {
  Container,
  Inner,
  Body,
  Frame,
  Header,
  Item,
  Title,
} from './styles/accordion';

type TAccordion = TRSC<typeof Container> & {
  Title: TRSC<typeof Title>;
  Frame: TRSC<typeof Frame>;
  Item: TRSC<typeof Item>;
  Header: TRSC<typeof Header>;
  Body: TRSC<typeof Body>;
};

const Accordion: TAccordion = ({ children, ...rest }) => {
  return (
    <Container {...rest}>
      <Inner>{children}</Inner>
    </Container>
  );
};

const ToggleContext = createContext<{
  show?: boolean;
  setShow?: Dispatch<SetStateAction<boolean>>;
}>({});

Accordion.Title = ({ children, ...rest }) => (
  <Title {...rest}>{children}</Title>
);

Accordion.Frame = ({ children, ...rest }) => (
  <Frame {...rest}>{children}</Frame>
);

Accordion.Item = function AccordionItem({ children, ...rest }) {
  const [show, setShow] = useState(false);

  return (
    <ToggleContext.Provider value={{ show, setShow }}>
      <Item {...rest}>{children}</Item>
    </ToggleContext.Provider>
  );
};

Accordion.Header = function AccordionHeader({ children, ...rest }) {
  const { show, setShow } = useContext(ToggleContext);

  const toggleShow = () => {
    setShow && setShow((show: boolean) => !show);
  };

  return (
    <Header onClick={toggleShow} {...rest}>
      {children}{' '}
      {show ? (
        <img src='/images/icons/close-slim.png' alt='' />
      ) : (
        <img src='/images/icons/add.png' alt='' />
      )}
    </Header>
  );
};

Accordion.Body = function AccordionBody({ children, ...rest }) {
  const { show } = useContext(ToggleContext);
  return (
    <Body className={show ? 'open' : 'closed'}>
      <span {...rest}>{children}</span>
    </Body>
  );
};

export default Accordion;
