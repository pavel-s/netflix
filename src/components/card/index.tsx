import { TRSC } from '../../types';

import {
  Container,
  CardContainer,
  Group,
  Title,
  Items,
  Item,
  ItemImage,
  Meta,
  Subtitle,
  Text,
  Feature,
  Maturity,
  FeatureClose,
} from './styles/card';

type TCard = TRSC<typeof Container> & {
  Container: TRSC<typeof Container>;
  Group: TRSC<typeof Group>;
  Title: TRSC<typeof Title>;
  Items: TRSC<typeof Items>;
  Item: TRSC<typeof Item>;
  ItemImage: TRSC<typeof ItemImage>;
  Meta: TRSC<typeof Meta>;
  Subtitle: TRSC<typeof Subtitle>;
  Text: TRSC<typeof Text>;
  Feature: TRSC<typeof Feature>;
  Maturity: TRSC<typeof Maturity>;
  FeatureClose: TRSC<typeof FeatureClose>;
};

const Card: TCard = ({ children, ...rest }) => {
  return <CardContainer {...rest}>{children}</CardContainer>;
};

Card.Container = function CardCardContainer({ children, ...rest }) {
  return <Container {...rest}>{children}</Container>;
};

Card.Group = function CardGroup({ children, ...rest }) {
  return <Group {...rest}>{children}</Group>;
};

Card.Title = function CardTitle({ children, ...rest }) {
  return <Title {...rest}>{children}</Title>;
};

Card.Items = function CardItems({ children, ...rest }) {
  return <Items {...rest}>{children}</Items>;
};

Card.Item = function CardItem({ children, ...rest }) {
  return <Item {...rest}>{children}</Item>;
};

Card.ItemImage = function CardItemImage({ children, ...rest }) {
  return <ItemImage {...rest}>{children}</ItemImage>;
};

Card.Meta = function CardMeta({ children, ...rest }) {
  return <Meta {...rest}>{children}</Meta>;
};

Card.Subtitle = function CardSubtitle({ children, ...rest }) {
  return <Subtitle {...rest}>{children}</Subtitle>;
};

Card.Text = function CardText({ children, ...rest }) {
  return <Text {...rest}>{children}</Text>;
};

Card.Feature = function CardFeature({ children, ...rest }) {
  return <Feature {...rest}>{children}</Feature>;
};

Card.Maturity = function CardMaturity({ maturity, ...rest }) {
  return (
    <Maturity maturity={maturity} {...rest}>
      {maturity}
    </Maturity>
  );
};

Card.FeatureClose = function CardFeatureClose({ children, ...rest }) {
  return <FeatureClose {...rest}>{children}</FeatureClose>;
};

export default Card;
