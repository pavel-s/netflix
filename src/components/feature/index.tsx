import { TRSC } from '../../types';
import { Container, Inner, Title, Subtitle } from './styles/feature';

type TFeature = TRSC<'div'> & {
  Title: TRSC<'h1'>;
  Subtitle: TRSC<'h2'>;
};

const Feature: TFeature = ({ children, ...rest }) => {
  return (
    <Container {...rest}>
      <Inner>{children}</Inner>
    </Container>
  );
};

Feature.Title = function FeatureTitle({ children, ...rest }) {
  return <Title {...rest}>{children}</Title>;
};

Feature.Subtitle = function FeatureSubtitle({ children, ...rest }) {
  return <Subtitle {...rest}>{children}</Subtitle>;
};

export default Feature;
