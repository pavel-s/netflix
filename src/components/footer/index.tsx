import { TRSC } from '../../types';
import { Container, Inner, Title, Links, Link, Text } from './styles/footer';

type TFooter = TRSC<'div'> & {
  Title: TRSC<'p'>;
  Links: TRSC<'ul'>;
  Link: TRSC<'a'>;
  Text: TRSC<'p'>;
};

const Footer: TFooter = ({ children, ...rest }) => {
  return (
    <Container {...rest}>
      <Inner>{children}</Inner>
    </Container>
  );
};

Footer.Title = ({ children, ...rest }) => {
  return <Title {...rest}>{children}</Title>;
};

Footer.Links = ({ children, ...rest }) => {
  return <Links {...rest}>{children}</Links>;
};

Footer.Link = ({ children, ...rest }) => {
  return (
    <Link>
      <a {...rest}>{children}</a>
    </Link>
  );
};

Footer.Text = ({ children, ...rest }) => {
  return <Text {...rest}>{children}</Text>;
};

export default Footer;
