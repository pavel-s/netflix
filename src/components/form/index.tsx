import { TRSC } from '../../types';
import {
  Container,
  Base,
  Title,
  Input,
  Error,
  Submit,
  Row,
  Checkbox,
  Text,
  TextBig,
  Link,
  TextButton,
} from './styles/form';

type TForm = TRSC<typeof Container> & {
  Base: TRSC<typeof Base>;
  Title: TRSC<typeof Title>;
  Input: TRSC<typeof Input>;
  Error: TRSC<typeof Error>;
  Submit: TRSC<typeof Submit>;
  Row: TRSC<typeof Row>;
  Checkbox: TRSC<typeof Checkbox>;
  Text: TRSC<typeof Text>;
  TextBig: TRSC<typeof TextBig>;
  Link: TRSC<typeof Link>;
  TextButton: TRSC<typeof TextButton>;
};

const Form: TForm = ({ children, ...rest }) => {
  return <Container {...rest}>{children}</Container>;
};

Form.Base = function FormBase({ children, ...rest }) {
  return <Base {...rest}>{children}</Base>;
};

Form.Title = function FormTitle({ children, ...rest }) {
  return <Title {...rest}>{children}</Title>;
};

Form.Input = function FormInput({ ...rest }) {
  return <Input {...rest} />;
};

Form.Error = function FormError({ ...rest }) {
  return <Error {...rest} />;
};

Form.Submit = function FormSubmit({ isLoading = false, children, ...rest }) {
  return (
    <Submit {...rest} isLoading={isLoading} type='submit'>
      {children}
    </Submit>
  );
};

Form.Row = function FormRow({ children, ...rest }) {
  return <Row {...rest}>{children}</Row>;
};

Form.Checkbox = function FormCheckbox({ children, ...rest }) {
  return <Checkbox {...rest}>{children}</Checkbox>;
};

Form.Text = function FormText({ children, ...rest }) {
  return <Text {...rest}>{children}</Text>;
};

Form.TextBig = function FormTextBig({ children, ...rest }) {
  return <TextBig {...rest}>{children}</TextBig>;
};

Form.Link = function FormLink({ children, ...rest }) {
  return <Link {...rest}>{children}</Link>;
};

Form.TextButton = function FormTextButton({ children, ...rest }) {
  return <TextButton {...rest}>{children}</TextButton>;
};

export default Form;
