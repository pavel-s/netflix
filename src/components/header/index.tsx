import { Background, Container, Logo, SignButton } from './styles/header';
import { LinkProps } from 'react-router-dom';
import { TRSC } from '../../types';

type THeader = TRSC<'div'> & {
  // Container: TRSC<'div'>;
  Logo: TRSC<'img'>;
  Frame: TRSC<'div'>;
  SignButton: TRSC<'a', LinkProps>;
};

const Header: THeader = ({ children, ...rest }) => {
  return <Background {...rest}>{children}</Background>;
};

Header.Frame = function HeaderFrame({ children, ...rest }) {
  return <Container {...rest}>{children}</Container>;
};

Header.Logo = function HeaderLogo({ children, ...rest }) {
  return <Logo {...rest} />;
};

Header.SignButton = function HeaderSignButton({ children, ...rest }) {
  return <SignButton {...rest}>{children}</SignButton>;
};

export default Header;
