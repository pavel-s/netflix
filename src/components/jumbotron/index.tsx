import { TRSC } from '../../types';
import {
  Container,
  Item,
  Inner,
  Pane,
  Title,
  Subtitle,
  Image,
  TFlexDirection,
} from './styles/jumbotron';

type TJumbotron = TRSC<typeof Item, { direction: TFlexDirection }> & {
  Container: TRSC<typeof Container>;
  Pane: TRSC<typeof Pane>;
  Title: TRSC<typeof Title>;
  Subtitle: TRSC<typeof Subtitle>;
  Image: TRSC<typeof Image>;
};

const Jumbotron: TJumbotron = ({ direction = 'row', children, ...rest }) => {
  return (
    <Item {...rest}>
      <Inner direction={direction}>{children}</Inner>
    </Item>
  );
};

Jumbotron.Container = function JumbotronContainer({ children, ...rest }) {
  return <Container {...rest}>{children}</Container>;
};

Jumbotron.Pane = function JumbotronPane({ direction, children, ...rest }) {
  return (
    <Pane direction={direction} {...rest}>
      {children}
    </Pane>
  );
};

Jumbotron.Title = function JumbotronTitle({ children, ...rest }) {
  return <Title {...rest}>{children}</Title>;
};

Jumbotron.Subtitle = function JumbotronSubtitle({ children, ...rest }) {
  return <Subtitle {...rest}>{children}</Subtitle>;
};

Jumbotron.Image = function JumbotronImage({ children, ...rest }) {
  return (
    <div>
      <Image {...rest} />
    </div>
  );
};

export default Jumbotron;
