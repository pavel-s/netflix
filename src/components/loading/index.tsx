import React from 'react';
import { TRSC } from '../../types';
import { Container, Image, LockBody, Spinner } from './styles/loading';

type TLoading = TRSC<
  typeof Image,
  Pick<typeof Spinner extends React.ComponentType<infer P> ? P : {}, 'size'>
>;

const Loading: TLoading = ({ size, src, ...rest }) => {
  return (
    <Container data-testid='loading'>
      <LockBody />
      <Spinner size={size}>
        {src ? <Image src={src} {...rest} /> : null}
      </Spinner>
    </Container>
  );
};

export default Loading;
