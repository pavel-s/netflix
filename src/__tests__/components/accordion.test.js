import { fireEvent, render } from '@testing-library/react';
import { Accordion } from '../../components';
import faqData from './../../fixtures/faqs.json';

describe('<Accordion />', () => {
  it('renders Accordion with FAQ data', () => {
    const { container, getByText } = render(
      <Accordion>
        <Accordion.Title>Frequently Asked Questions</Accordion.Title>
        <Accordion.Frame>
          {faqData.map((item) => (
            <Accordion.Item key={item.id}>
              <Accordion.Header>{item.header}</Accordion.Header>
              <Accordion.Body>{item.body}</Accordion.Body>
            </Accordion.Item>
          ))}
        </Accordion.Frame>
      </Accordion>
    );

    // accordion title
    expect(getByText('Frequently Asked Questions')).toBeTruthy();

    // item header
    const header = getByText(faqData[1].header);
    expect(header).toBeTruthy();

    // item body hidden
    const body = getByText(faqData[1].body).parentElement;
    expect(body).toBeTruthy();
    expect(body).toHaveClass('closed');

    // header with + icon
    expect(header.lastChild).toHaveAttribute('src', '/images/icons/add.png');

    // click on header, icon changed, body open
    fireEvent.click(header);
    expect(header.lastChild).toHaveAttribute(
      'src',
      '/images/icons/close-slim.png'
    );
    expect(body).toHaveClass('open');

    // click again
    fireEvent.click(header);
    expect(header.lastChild).toHaveAttribute('src', '/images/icons/add.png');
    expect(body).toHaveClass('closed');

    expect(container.firstChild).toMatchSnapshot();
  });
});
