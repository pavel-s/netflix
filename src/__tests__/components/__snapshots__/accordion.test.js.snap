// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Accordion /> renders Accordion with FAQ data 1`] = `
<div
  class="accordion__Container-sc-1pxz54l-0 dsionx"
>
  <div
    class="accordion__Inner-sc-1pxz54l-1"
  >
    <h1
      class="accordion__Title-sc-1pxz54l-2 dbluKp"
    >
      Frequently Asked Questions
    </h1>
    <div
      class="accordion__Frame-sc-1pxz54l-3 cUHtCp"
    >
      <div
        class="accordion__Item-sc-1pxz54l-4 cImjvy"
      >
        <div
          class="accordion__Header-sc-1pxz54l-5 hRaHcY"
        >
          What is Netflix?
           
          <img
            alt=""
            src="/images/icons/add.png"
          />
        </div>
        <div
          class="accordion__Body-sc-1pxz54l-6 hWJBhX closed"
        >
          <span>
            Netflix is a streaming service that offers a wide variety of award-winning TV shows, movies, anime, documentaries, and more on thousands of internet-connected devices.

You can watch as much as you want, whenever you want without a single commercial – all for one low monthly price. There's always something new to discover and new TV shows and movies are added every week!
          </span>
        </div>
      </div>
      <div
        class="accordion__Item-sc-1pxz54l-4 cImjvy"
      >
        <div
          class="accordion__Header-sc-1pxz54l-5 hRaHcY"
        >
          How much does Netflix cost?
           
          <img
            alt=""
            src="/images/icons/add.png"
          />
        </div>
        <div
          class="accordion__Body-sc-1pxz54l-6 hWJBhX closed"
        >
          <span>
            Watch Netflix on your smartphone, tablet, Smart TV, laptop, or streaming device, all for one fixed monthly fee. Plans range from RUB599 to RUB999 a month. No extra costs, no contracts.
          </span>
        </div>
      </div>
      <div
        class="accordion__Item-sc-1pxz54l-4 cImjvy"
      >
        <div
          class="accordion__Header-sc-1pxz54l-5 hRaHcY"
        >
          Where can I watch?
           
          <img
            alt=""
            src="/images/icons/add.png"
          />
        </div>
        <div
          class="accordion__Body-sc-1pxz54l-6 hWJBhX closed"
        >
          <span>
            Watch anywhere, anytime, on an unlimited number of devices. Sign in with your Netflix account to watch instantly on the web at netflix.com from your personal computer or on any internet-connected device that offers the Netflix app, including smart TVs, smartphones, tablets, streaming media players and game consoles.

You can also download your favorite shows with the iOS, Android, or Windows 10 app. Use downloads to watch while you're on the go and without an internet connection. Take Netflix with you anywhere.
          </span>
        </div>
      </div>
      <div
        class="accordion__Item-sc-1pxz54l-4 cImjvy"
      >
        <div
          class="accordion__Header-sc-1pxz54l-5 hRaHcY"
        >
          How do I cancel?
           
          <img
            alt=""
            src="/images/icons/add.png"
          />
        </div>
        <div
          class="accordion__Body-sc-1pxz54l-6 hWJBhX closed"
        >
          <span>
            Netflix is flexible. There are no pesky contracts and no commitments. You can easily cancel your account online in two clicks. There are no cancellation fees – start or stop your account anytime.
          </span>
        </div>
      </div>
      <div
        class="accordion__Item-sc-1pxz54l-4 cImjvy"
      >
        <div
          class="accordion__Header-sc-1pxz54l-5 hRaHcY"
        >
          What can I watch on Netflix?
           
          <img
            alt=""
            src="/images/icons/add.png"
          />
        </div>
        <div
          class="accordion__Body-sc-1pxz54l-6 hWJBhX closed"
        >
          <span>
            Netflix has an extensive library of feature films, documentaries, TV shows, anime, award-winning Netflix originals, and more. Watch as much as you want, anytime you want.
          </span>
        </div>
      </div>
      <div
        class="accordion__Item-sc-1pxz54l-4 cImjvy"
      >
        <div
          class="accordion__Header-sc-1pxz54l-5 hRaHcY"
        >
          Is Netflix good for kids?
           
          <img
            alt=""
            src="/images/icons/add.png"
          />
        </div>
        <div
          class="accordion__Body-sc-1pxz54l-6 hWJBhX closed"
        >
          <span>
            The Netflix Kids experience is included in your membership to give parents control while kids enjoy family-friendly TV shows and movies in their own space.

Kids profiles come with PIN-protected parental controls that let you restrict the maturity rating of content kids can watch and block specific titles you don’t want kids to see.
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
